src/sys/signalfd.rs

```diff
@@ -82,30 +85,30 @@ pub fn signalfd(fd: RawFd, mask: &SigSet, flags: SfdFlags) -> Result<RawFd> {
 ///     Err(err) => (), // some error happend
 /// }
 /// ```
-#[derive(Debug, Eq, Hash, PartialEq)]
-pub struct SignalFd(RawFd);
+#[derive(Debug)]
+pub struct SignalFd(ManuallyDrop<OwnedFd>);
```

Comment:
Why use `ManuallyDrop`?  An `OwnedFd` will close itself on drop.  Relying on that would simplify things.

Reply:
Previously the closing of the file descriptor was handled such that it would panic if `unistd::close` failed:
```rust
impl Drop for SignalFd {
    fn drop(&mut self) {
        let e = unistd::close(self.0.as_raw_fd());
        if !std::thread::panicking() && e == Err(Errno::EBADF) {
            panic!("Closing an invalid file descriptor!");
        };
    }
}
```
Without `ManuallyDrop` we rely upon how the [standard library `Drop` implementation](https://doc.rust-lang.org/stable/src/std/os/fd/owned.rs.html#167-179) which simply ignore the return value of `close` meaning it would no longer panic if it failed.
```rust
impl Drop for OwnedFd {
    #[inline]
    fn drop(&mut self) {
        unsafe {
            // Note that errors are ignored when closing a file descriptor. The
            // reason for this is that if an error occurs we don't actually know if
            // the file descriptor was closed or not, and if we retried (for
            // something like EINTR), we might close another valid file descriptor
            // opened after we closed ours.
            let _ = libc::close(self.fd);
        }
    }
}
```

Reply:
By definition, it is UB for an OwnedFd to use an invalid file descriptor, so that code can be removed.

Reply:
Is it technically UB?  I think the compiler will be perfectly happy with it.  The memory is initialized, after all.